Feature: Implement Sign-Up Rate Limiting

ben-fornefeld opened the pull request:
This pr adds an ip based sign-up counter using Redis. all crucial values are configurable via environment variables. 

The configuration for production is currently set to:
```bash
  SIGN_UP_LIMIT_PER_WINDOW=1
  SIGN_UP_WINDOW_HOURS=24
```

If the user has too many sign-up's, he might see rate limit alert like this:

<img width="417" height="671" alt="Screenshot 2025-09-09 at 1 53 30 PM" src="https://github.com/user-attachments/assets/ac8e7c5b-9e5b-46df-bd58-a69a42227200" />

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduces IP-based sign-up rate limiting using Vercel KV with configurable env vars and integrates checks into `signUpAction`.
> 
> - **Auth/Backend**:
>   - Add IP-based sign-up rate limiting in `src/server/auth/auth-actions.ts` using `ipAddress` and new helpers `incrementAndCheckSignUpRateLimit`/`decrementSignUpRateLimit`.
>   - New module `src/server/auth/ratelimit.ts` implements atomic Redis Lua scripts for increment/decrement and TTL handling.
>   - Extend error handling (`email_address_invalid`) and warn when no IP in production.
> - **Config/Env**:
>   - New flag `ENABLE_SIGN_UP_RATE_LIMITING` in `src/configs/flags.ts`; add KV-backed `KV_KEYS.RATE_LIMIT_SIGN_UP` and defaults in `src/configs/limits.ts`.
>   - Expand `src/lib/env.ts` schemas: add numeric-boolean support and new server vars `ENABLE_SIGN_UP_RATE_LIMITING`, `SIGN_UP_LIMIT_PER_WINDOW`, `SIGN_UP_WINDOW_HOURS`; tighten client `NEXT_PUBLIC_*` booleans.
>   - `.env.example`: document optional Auth rate limiting envs.
> - **Dependencies/Build**:
>   - Add `@vercel/functions` dependency.
>   - `tsconfig.json`: set `jsx` to `preserve`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 0f9995749055b7bcd17a749e2a2951f30b235476. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

ben-fornefeld committed:
Feature: Implement Sign-Up Rate Limiting

ben-fornefeld committed:
patch: Update Sign-Up Rate Limiting Flag Logic

mishushakov commented on src/app/api/auth/confirm/route.ts:
```diff
@@ -135,6 +137,17 @@ export async function GET(request: NextRequest) {
+    // increment counter for successful sign-up confirmations (rate limiting)
+    if (ENABLE_SIGN_UP_RATE_LIMITING && supabaseType === 'signup') {
+      const ip =
+        request.headers.get('x-forwarded-for') ||
+        request.headers.get('cf-connecting-ip') ||
+        request.headers.get('x-real-ip') ||
+        'unknown'
```
in this condition if none of the headers are forwarded the usage will be counted towards 'unknown', so that means everyone will be affected by rate limit

ben-fornefeld replied:
i know. it should never happen but that is what the kill switch env is for. we get the errors logged in grafana.

mishushakov commented on src/server/auth/auth-actions.ts:
```diff
@@ -93,6 +95,18 @@ export const signUpAction = actionClient
     }
 
+    const ip =
+      (await headers()).get('x-forwarded-for') ||
+      (await headers()).get('cf-connecting-ip') ||
+      (await headers()).get('x-real-ip') ||
```
deduplicate await headers()

ben-fornefeld replied:
it does not await actually. this is only for the next framework to handle this file dynamically. nextjs uses await to signal this to the framework.

ben-fornefeld committed:
remove: manual reset since ttl resets

mishushakov commented on src/server/auth/rate-limiting.ts:
```diff
@@ -0,0 +1,74 @@
+    const key = KV_KEYS.SIGN_UP_RATE_LIMIT(identifier)
+    const current = await kv.get(key)
+    const currentCount = current ? parseInt(current as string, 10) : 0
+    const newCount = currentCount + 1
```
please check the the value is not negative, otherwise I might be able to pass -100

ben-fornefeld replied:
how can it be negative, there is no decrement anywhere in the codebase?

mishushakov replied:
please increment in Redis using INCR, so you won't need kv.get anymore

jakubno replied:
race condition, if you are updating concurrently on multiple places, you will get on all of them 0 and the increase to 1, this can happen n times, so the counter will be 1 after n concurrent requests

ben-fornefeld committed:
feat: Add Upstash Rate Limiting for Sign-Up Process

- Introduced `@upstash/ratelimit` dependency for improved rate limiting functionality.
- Replaced manual increment logic with Upstash's rate limiting utility in the sign-up process.
- Updated sign-up rate limiting checks to enhance logging and error handling.
- Refactored related functions to streamline the rate limiting implementation.

ben-fornefeld committed:
refactor: Reorganize OTP verification logic in auth confirmation route

ben-fornefeld committed:
refactor: Update rate limiting logic and environment schema for sign-up process

- Renamed `TruethyOrFalsy` to `NumericBoolean` for clarity in environment schema.
- Added new fields for sign-up attempts limit and window in the environment schema.
- Updated rate limiting functions to improve clarity and logging for sign-up attempts.
- Refactored related code to enhance consistency in rate limiting checks.

ben-fornefeld committed:
refactor: Update environment variables and rate limiting configuration for sign-up process

- Renamed `SIGN_UP_ATTEMPTS_LIMIT` to `SIGN_UP_ATTEMPTS_LIMIT_PER_WINDOW` for clarity.
- Added new environment variables for sign-up rate limiting configuration.
- Updated documentation and related code to reflect the changes in rate limiting logic.

ben-fornefeld committed:
fix: Validate IP address in auth confirmation and sign-up actions

- Added checks to ensure a valid IP address is retrieved in both the auth confirmation route and sign-up action.
- Returns an error response if the IP address is invalid, improving error handling and user feedback.

ben-fornefeld committed:
refactor: Enhance rate limiting configuration with positive number validation

- Introduced a custom string type for validating positive numbers in the environment schema.
- Updated rate limiting configuration to ensure that limits and windows are positive numbers.
- Improved documentation to reflect the new validation requirements for rate limiting parameters.

ben-fornefeld committed:
refactor: Improve IP address handling and logging in rate limiting

- Refactored IP address extraction in the auth confirmation route and sign-up action to use a dedicated utility function.
- Enhanced logging to capture errors when IP headers are missing in production environments.
- Updated documentation to clarify the handling of IP addresses in the rate limiting system, including support for development environments.

ben-fornefeld commented:
> Did you test it on staging?

@mishushakov tested on my cluster.

ben-fornefeld committed:
Merge branch 'main' into feat-sign-up-request-limiting-per-ip-address-e2b-2967

ben-fornefeld committed:
refactor: Simplify rate limiting logic and remove deprecated configurations

ben-fornefeld committed:
refactor: Remove rate limiting logic from sign-up confirmation route

ben-fornefeld committed:
refactor: rate limit handling to simplify

ben-fornefeld committed:
chore: clean up

ben-fornefeld committed:
wip: fix race condition

ben-fornefeld committed:
remove: rate limit lib in favor of incr/decr

ben-fornefeld commented:
bugbot run

ben-fornefeld committed:
refactor: use custom lua script to ensure atomic mutations

ben-fornefeld commented:
bugbot run

ben-fornefeld committed:
fix: missing await

ben-fornefeld commented:
bugbot run

ben-fornefeld commented on src/server/auth/auth-actions.ts:
```diff
@@ -103,6 +113,40 @@ export const signUpAction = actionClient
+      return returnServerError(
+        'Too many sign-up attempts. Please try again later.'
+      )
+    }
```
it's expected behavior here.

ben-fornefeld committed:
fix: ensure correct number passing + ensure ttl is always set

ben-fornefeld committed:
refactor: streamline sign-up rate limiting by using default values and simplifying logic

ben-fornefeld commented:
closing this because it's stale and not aligned with the current roadmap anymore.
